### PR TITLE
MDL-79460: Add support for powerlevels and route for sync

### PR DESCRIPTION
**Pull request by davewoloszyn**
I have added a new route for /v3/sync. This commit allows certain behat and phpunit tests to function again. See MDL-79460 for more info.

There is also some power level approximation going on in there for students, teachers and admins. This will let Matrix assign them the equivalent power level that Moodle uses based on roles by checking the Matrix username used.

There is a noticeable downside to doing this. If a user who has the username of @teacher123, they will be assigned the powerlevel of 51, but if they are demoted, or even promoted, their powerlevel will be set back to what the username indicates it should be.

### CODE REVIEW

**Review by andrewnicols — CHANGES_REQUESTED** [review]
Hi David,

Im' not sure I understand this change.

When a room is created it will have an empty power configuration with just an initial administrator (the creator).

The event to set the power level shoudl be setting these values, which we then fetch back.

We need to store these somewhere in the DB, which means adding a DB field for the JSON. We then just need to store and retrieve that exact value. We should not have any knowledge of the Moodle moderator, or site admin, nor any moodle roles, nor guessing a power level based upon the username.

Basically we need to:
- introduce a new handler for the existing event route, which handles m.room.power_levels and stores the values provided as JSON
- introduce the /sync route, and have it return this information
- hav a default value if none has been set yet

**Comment by davewoloszyn** [comment]
Thanks Andrew for the feedback. I think I understand the mock server a little better now.

1. I have removed all Moodle references and power level approximations
2. New column for RoomMember can keep track of a user's power level
3. Default (0) is set if no power level is set

Let me know if there is anything else I can improve upon. Thanks.

File: application/src/Controller/MatrixController.php [hunk 1]
```diff
@@ -402,6 +488,29 @@ public function roomState(string $serverID, string $roomID, string $eventType, R
...
+                if (!empty($level)) {
+                    $user = $entityManager->getRepository(User::class)->findOneBy([
+                        'serverid' => $serverID,
+                        'userid' => $userid,
+                    ]);
+                    if (!empty($user)) {
+                        // Don't modify admin power levels.
+                        if ($user->getAdmin()) {
+                            break;
+                        }
```
Comment (andrewnicols) [inline application/src/Controller/MatrixController.php]: Remove - there are no preventions in the API for this.

File: application/src/Controller/MatrixController.php [hunk 2]
```diff
@@ -273,6 +273,92 @@ public function createRoom(string $serverID, Request $request):JsonResponse {
         return new JsonResponse((object) $response, 200);
     }
 
+    /**
+     * Get room power levels.
+     *
+     * @Route("/v3/sync")
```
Comment (andrewnicols) [inline application/src/Controller/MatrixController.php:281]: I'd suggest posting a link to the relevant part of the Matrix spec for this.

File: application/src/Controller/MatrixController.php [hunk 3]
```diff
@@ -273,6 +273,92 @@ public function createRoom(string $serverID, Request $request):JsonResponse {
...
+     * @param string $serverID
+     * @param Request $request
+     * @return JsonResponse
+     */
+    public function getRoomPowerLevels(string $serverID, Request $request):JsonResponse {
```
Comment (andrewnicols) [inline application/src/Controller/MatrixController.php]: this is really a sync request. At the moment we only fulfill this part but we may have to fulfill other parts later.

File: application/src/Entity/RoomMember.php [hunk 4]
```diff
@@ -154,4 +159,16 @@ public function setState(string $state = null): self
 
         return $this;
     }
+
+    public function getPowerlevel(): ?int
```
Comment (andrewnicols) [inline application/src/Entity/RoomMember.php]: Probably should be uppercase L in `getPowerLevel`

File: application/src/Controller/MatrixController.php [hunk 5]
```diff
@@ -273,6 +273,92 @@ public function createRoom(string $serverID, Request $request):JsonResponse {
...
+
+        // Power level values.
+        $powerLevelDefault = 0;
+        $powerLevelModerator = 50;
+        $powerLevelMaximum = 100;
+
+        $entityManager = $this->getDoctrine()->getManager();
+
+        // Room info.
+        $room = $entityManager->getRepository(Room::class)->findOneBy([
```
Comment (andrewnicols) [inline application/src/Controller/MatrixController.php]: How does it know which room to fetch? This will only fetch the first.

This request should fetch all rooms where the requesting user is a member.
In our case the requesting user is our admin user but I don't think we put the admin user in the room (we should, but separate issue to solve).

The request _may_ (does in our case) include a filter which we also need to honour.

Essentially this needs to be:

```
// Need to return a new 'Filter' object I think to help us parse the provided filter and turn it into something meaningful.
$filter = Filter::fromRequest($request);

$roomSearch = [
    'serverid' => $serverID,
];

if ($roomIds = $filter->getRoomFilter()->getRoomIds()) {
    // Doctrine will convert an array into a WHERE IN.
    $roomSearch['roomid'] = $roomIds;
}

$roomList = $entityManager->getRepository(Room::class)->findBy($roomSearch);

$roomData = [];
foreach ($roomList as $room) {
    $roomData[$room->getRoomId()] = [
        'state' => $room->getRoomState($request),
    ];
}

// ...
return new JsonResponse((object) [
    'rooms' = [
        'join' => $roomData,
    ],
]);
```

And add a new method to either the Entity:

```
namespace App\Entity;
class Room
{
    // ...
    public function getRoomState(Filter $filter): array {
        // We pass the Filter, but we don't actually use it yet.
        // We will probably need to in the future.
        // At the moment we only know how to serve a set of power levels.
        $roomData = [
            'state' => [
                'events' => [],
            ],
        ];

        $roomData['state']['events'][] = $this->getPowerLevelState();

        return $roomData;
    }

    protected function getPowerLevelState(): array {
        // Power level values.
        $powerLevelDefault = 0;
        $powerLevelModerator = 50;
        $powerLevelMaximum = 100;

        // Get the members of the room who have a powerlevel set.
        $members = $this->getMembers()->filter(function (RoomMember $member) {
            return $member->getPowerLevel() !== null;
        });

        // Build the power level state.
        // Note: Not 100% sure on this bit without testing it. This may return an array of arrays and need flattening and there may be a better way of doing it.
        $memberInfo = $members->map(fn(RoomMember $member) => [
            $member->getUser()->getUserid() => $member->getPowerLevel(),
        ])->toArray();

        return [
            'type' => 'm.room.power_levels',
            'content' => [
                'users' => (object) $memberInfo,
                'users_default' => $powerLevelDefault,
                'events' => [
                    'm.room.name' => $powerLevelModerator,
                    'm.room.power_levels' => $powerLevelMaximum,
                    'm.room.history_visibility' => $powerLevelMaximum,
                    'm.room.canonical_alias' => $powerLevelModerator,
                    'm.room.avatar' => $powerLevelModerator,
                    'm.room.tombstone' => $powerLevelMaximum,
                    'm.room.server_acl' => $powerLevelMaximum,
                    'm.room.encryption' => $powerLevelMaximum,
                ],
                'events_default' =>  $powerLevelDefault,
                'state_default' => $powerLevelModerator,
                'ban' => $powerLevelModerator,
                'kick' => $powerLevelModerator,
                'redact' => $powerLevelModerator,
                'invite' =>  $powerLevelDefault,
                'historical' => $powerLevelMaximum,
            ],
        ];
    }
}
```

And then the Filter class:
```
// Not sure of the best location for this. Maybe Service? Maybe Util?
namespace App\Service;

use App\Service\Filter\RoomFilter;
use Symfony\Component\HttpFoundation\Request;

class Filter {
    private array $roomIds = [];
    private RoomFilter $roomFilter;

    public static function fromRequest(Request $request): static
    {
        $filter = new static();

        $payload = $request->toArray();
        if (!array_key_exists('filter', $payload)) {
            return $filter;
        }

        $filterData = $payload['filter'];

        // Filter the event data according to the API:
        // https://spec.matrix.org/v1.1/client-server-api/#filtering
        // Filters are made up of different filter types.
        // Detailed here: https://spec.matrix.org/v1.1/client-server-api/#post_matrixclientv3useruseridfilter
        // We only need to implement a handful of these.
        if (array_key_exists('room', $filterData)) {
            $filter->setRoomFilter(RoomFilter::fromArray($filterData['room']));
        }

        return $filter;
    }

    public function setRoomFilter(RoomFilter $roomFilter): static
    {
        $this->roomFilter = $roomFilter;
        return $this;
    }

    public function getRequestedRoomIds(): array
    {
        return $this->roomIds;
    }

    public function getRoomFilter(): RoomFilter
    {
        if (empty($this->roomFilter)) {
            $this->roomFilter = new RoomFilter();
        }

        return $this->roomFilter;
    }
}
```

And then corresponding RoomFilter:
```
<?php

namespace App\Service\Filter;

class RoomFilter
{
    private array $roomIds = [];
    private StateFilter $roomStateFilter;

    public static function fromArray(array $filterData): static
    {
        $filter = new static();

        // Check for a list of rooms.
        if (array_key_exists('rooms', $filterData)) {
            $filter->setRoomIds($filterData['rooms']);
        }

        // Check for a room state filter.
        if (array_key_exists('state', $filterData)) {
            $filter->setRoomState(StateFilter::fromArray($filterData['state']));
        }

        return $filter;
    }

    public function setRoomIds(array $ids): static
    {
        $this->roomIds = $ids;
        return $this;
    }

    public function setRoomState(StateFilter $state): static
    {
        $this->roomStateFilter = $state;
        return $this;
    }

    public function getRoomIds(): array
    {
        return $this->roomIds;
    }

    public function getRoomState(): StateFilter
    {
        return $this->roomStateFilter;
    }
}

```

And StateFilter:
```
<?php

namespace App\Service\Filter;

class StateFilter
{
    private array $roomIds = [];
    private StateFilter $roomStateFilter;

    public static function fromArray(array $filterData): static
    {
        $filter = new static();

        // We currently only support types.
        if (array_key_exists('types', $filterData)) {
            $filter->setTypes($filterData['types']);
        }

        return $filter;
    }

    public function setTypes(array $types): static
    {
        $this->types = $types;
        return $this;
    }

    public function getTypes(): array
    {
        return $this->types;
    }
}
```

**Review by andrewnicols — CHANGES_REQUESTED** [review]

File: application/src/Controller/MatrixController.php (hunk 1 above)
Comment (davewoloszyn) [reply application/src/Controller/MatrixController.php]: I have removed this.

File: application/src/Controller/MatrixController.php (hunk 2 above)
Comment (davewoloszyn) [reply application/src/Controller/MatrixController.php:281]: I have added this.

File: application/src/Controller/MatrixController.php (hunk 3 above)
Comment (davewoloszyn) [reply application/src/Controller/MatrixController.php]: Yes, you are right and your approach has made this clear. I have renamed to 'sync'.

File: application/src/Entity/RoomMember.php (hunk 4 above)
Comment (davewoloszyn) [reply application/src/Entity/RoomMember.php]: Fixed.

File: application/src/Controller/MatrixController.php (hunk 5 above)
Comment (davewoloszyn) [reply application/src/Controller/MatrixController.php]: Thank you for providing this. It really helped.

I had to make some adjustments. Nothing major.

All communication tests are passing in Behat and PHPUnit with these changes.

**Review by andrewnicols — APPROVED** [review]
